### PR TITLE
Set travis to build docker images and test against them.

### PR DESCRIPTION
**Pull request by glogiotatidis**

### CODE REVIEW

**Comment by glogiotatidis** [comment]
Pinging both @pmclanahan and @jgmize for this new deis deployment approach.

I still want to run a couple test runs with travis to investigate docker caching possibilities but this is mostly done and I'd love your feedback. I've deployed to deis and it works! :)

File: .travis.yml [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
-  - TOX_ENV=flake8
-  - TOX_ENV=docs
-  - TOX_ENV=tests
+  global:
+    # NewRelic
+    - NEWRELIC_STAGE_APPLICATION=careers-stage-us-west
+    - NEWRELIC_PROD_APPLICATION=careers-prod-us-west
+    # Deis
+    - DEIS_STAGE_APPLICATION=careers-stage
+    - DEIS_PROD_APPLICATION=careers-stage
```
Comment (pmclanahan) [inline .travis.yml]: should this be `careers-prod`?

File: careers/careers/utils.py [hunk 2]
```diff
@@ -14,10 +14,10 @@ def order_by_function(x):
         order_by = order_by_function
 
     return (
-        sorted(set(
+        sorted(
```
Comment (pmclanahan) [inline careers/careers/utils.py]: Should this change and the one for `test_utils.py` be a part of this PR? Seems unrelated.

File: .travis.yml [hunk 3]
```diff
@@ -1,31 +1,46 @@
...
+    - DOCKER_EMAIL="foo@example.com"
+    - DOCKER_USERNAME="mozjenkins"
+    # Django
+    - DEBUG=False
+    - ALLOWED_HOSTS=*
+    - SECRET_KEY=foo
+    - DATABASE_URL=postgres://postgres@db/postgres
+install:
+  - docker pull ${DOCKER_REPOSITORY}:last_successful_build || true
+  - docker pull ${DOCKER_REPOSITORY}:${TRAVIS_COMMIT} || true
```
Comment (jgmize) [inline .travis.yml]: why are the `|| true`s needed here?

**Comment by jgmize** [comment]
Nice work, @glogiotatidis. This looks like a promising approach. One thing I'd like to see in the future (not necessarily part of this PR) is the addition of some integration tests that get triggered after a successful deploy.

File: .travis.yml (hunk 1 above)
Comment (glogiotatidis) [inline .travis.yml]: good catch, thnx!

File: .travis.yml (hunk 3 above)
Comment (glogiotatidis) [inline .travis.yml]: So here I'm trying to take advantage of docker caching. 

* Before deploying I'm pushing to docker hub with the last_successful_build tag and the git sha tag. 
* During the next travis run (let's say I'm tagging the commit and I'm pushing to production) these lines will download the already built image and docker will skip re-building it and use the cache.
* The `|| true` part of the cmds is there in case there is no `last_successful_build` tag or a tag with the current tested SHA commit. In this case the docker client will return a non zero exit code and travis will stop running the tests when in fact this is not a blocker. Thus I'm adding `|| true` in the end of the line to make sure that the cmd returns successfully regardless docker pull  

For example you can see these sequential runs where I changed `requirements.txt`. The docker layers before `peep install` are all cached.

https://s3.amazonaws.com/archive.travis-ci.org/jobs/103077087/log.txt
https://s3.amazonaws.com/archive.travis-ci.org/jobs/103080065/log.txt

**Comment by glogiotatidis** [comment]
> Nice work, @glogiotatidis. This looks like a promising approach. One thing I'd like to see in the future (not necessarily part of this PR) is the addition of some integration tests that get triggered after a successful deploy.

In my todo list. I need to find a (non-hacky) way to make a travis pipeline much like we have for bedrock.

File: careers/careers/utils.py (hunk 2 above)
Comment (glogiotatidis) [inline careers/careers/utils.py]: Fair enough. This ended up here b/c while testing this PR I got failing tests due to lists containing the same items but with different order. But indeed this has nothing to do with the travis changes. 

I opt to keep this in the same PR to avoid failing tests but I moved it into its separate commit.

Comment (glogiotatidis) [inline careers/careers/utils.py]: And now with more caching using `ADD`. See https://github.com/docker/docker/pull/12031

**Comment by jgmize** [comment]
> And now with more caching using ADD. See docker/docker#12031

My reading of both the discussion and the docs in the PR is that that both ADD and COPY now ignore mtime: https://github.com/docker/docker/pull/12031/files#diff-9fa9788827578dc424e6de362de15e3cL102

According to https://docs.docker.com/engine/articles/dockerfile_best-practices/#add-or-copy COPY is preferred in all situations where you don't need to fetch a URL or extract a tarball.

**Comment by glogiotatidis** [comment]
> My reading of both the discussion and the docs in the PR is that that both ADD and COPY now ignore mtime: https://github.com/docker/docker/pull/12031/files#diff-9fa9788827578dc424e6de362de15e3cL102

 I was sure that I tried it and it didn't work with COPY and I was under that impression that it was a docker build bug. I tried again it works in both cases. I reverted to using COPY again.


First with ADD
https://travis-ci.org/mozilla/lumbergh/builds/103320779
https://api.travis-ci.org/jobs/103320780/log.txt?deansi=true

Second with ADD
https://travis-ci.org/mozilla/lumbergh/builds/103321598
https://s3.amazonaws.com/archive.travis-ci.org/jobs/103321599/log.txt

First with COPY
https://travis-ci.org/mozilla/lumbergh/builds/103323241
https://api.travis-ci.org/jobs/103323242/log.txt?deansi=true


Second with COPY
https://travis-ci.org/mozilla/lumbergh/builds/103324515
https://s3.amazonaws.com/archive.travis-ci.org/jobs/103324516/log.txt

> According to https://docs.docker.com/engine/articles/dockerfile_best-practices/#add-or-copy COPY is preferred in all situations where you don't need to fetch a URL or extract a tarball.

You're right. That's why I had COPY in the first place.

**Comment by glogiotatidis** [comment]
Hey folks, this is ready from my side. Good to merge?

**Comment by pmclanahan** [comment]
LGTM @glogiotatidis. Merge at will. :dancers:
